nixos/modules/services/backup/bupstash.nix

```diff
@@ -0,0 +1,281 @@
+                default = "root";
+                description = "User for the bupstash server.";
+              };
+
+              group = mkOption {
+                type = str;
+                default = "root";
+                description = "Group for the bupstash server.";
+              };
+
+              preStash = mkOption {
+                type = lines;
+                default = "";
+                description = "Commands to run before `bupstash put`.";
+              };
+
+              postStash = mkOption {
+                type = lines;
+                default = "";
+                description = "Commands to run after `bupstash put`.";
+              };
+
+              repositories = mkOption {
+                type = listOf (
+                  either str (
+                    submodule {
+                      options = {
+                        user = mkOption {
+                          type = str;
+                          description = "User used to connect to the remote server.";
+                        };
+                        server = mkOption {
+                          type = str;
+                          description = "Address ot the remote server.";
+                        };
+                        path = mkOption {
+                          type = nullOr str;
+                          description = "Remote path for the repository.";
+                        };
+                      };
+                    }
+                  )
+                );
+                default = [ ];
+                description = "List of repositories to backup to.";
+              };
+
+              tags = mkOption {
+                type = attrsOf str;
+                default = {
+                  inherit name;
+                };
+                description = "List of tags for this job.";
+              };
+
+              startAt = mkOption {
+                type = either str (listOf str);
+                default = [ ];
+                description = ''
+                  Automatically start this backup at the given date/time, which
+                  must be in the format described in {manpage}`systemd.time(7)`.
+                '';
+              };
+
+              extraOptions = mkOption {
+                type = listOf str;
+                default = [ ];
+                description = ''
+                  Additional options passed to `bupstash put`.
+                '';
+              };
+            };
+
+            config = {
+              extraOptions = builtins.map (pattern: "--exclude '${pattern}'") config.excludes;
+            };
+          }
+        )
+      );
+      default = { };
+      description = "List of backup jobs to run using bupstash.";
+    };
+  };
+
+  config = mkMerge [
+    (mkIf (cfg.jobs != {}) {
+      systemd.tmpfiles.rules =
+        mapAttrsToList
+          (name: job: "d /var/lib/bupstash/send-logs/${name} 0700 ${job.user} ${job.group} - -")
+          cfg.jobs;
```

Comment:
Given that bupstash owns its working directory, can it not create that path itself? It really should, and if it does not, consider creating an upstream issue, so we can eventually remove this hack.

Are you worried about the permissions on that path? Either way, please set `StateDirectoryMode=0700` and consider `UMask = "0077"`.

Reply:
After testing, bupstash does not create this path alone, it can create the file for the repository log, but does not create the parent directories if they do not exist

Reply:
After some more research, it is tied to the call to `sqlite3_open_v2` inside rusqlite, so it might be hard to fix upstream

Reply:
Just setting `environment.HOME = "/var/lib/bupstash"` should solve this.

> --send-log PATH
>
>    Path to the send log file, defaults to one of the following, in order, provided the appropriate environment variables are set, $BUPSTASH_SEND_LOG, $XDG_CACHE_HOME/.cache/bupstash/bupstash.sendlog or $HOME/.cache/bupstash/bupstash.sendlog.

https://bupstash.io/doc/man/bupstash-put.html

Reply:
From the manpages, 
> The send log only remembers the data previously sent, so for efficient 'put' use, give each backup job a unique send log file.
We would need different `$HOME` for each pair of (`job`,`repository`)

Reply:
When splitting the job into multiple units, the dynamicuser meant that creating state directories for each repo in each job was necessary anyway

Reply:
Bupstash should really take the repository into account when creating send logs. I think we should file an upstream issue for this.